symforce/slam/imu_preintegration/manifold_symbolic.py

```diff
@@ -0,0 +1,346 @@
+            in the body frame at timestep i: R_i^T (p_j - p_i - v_i Delta t - 1/2 g Delta t^2),
+            where R_i = pose_i.R, p_i = pose_i.t, and p_j = pose_j.t
+        sqrt_info (sf.M99): sqrt info matrix of DR('s tangent space), Dv, Dp
+        DR_D_gyro_bias (sf.M33): Derivative of DR w.r.t. gyro_bias evaluated at gyro_bias_hat
+        Dp_D_accel_bias (sf.M33): Derivative of Dp w.r.t. accel_bias evaluated at accel_bias_hat
+        Dp_D_gyro_bias (sf.M33): Derivative of Dp w.r.t. gyro_bias evaluated at gyro_bias_hat
+        Dv_D_accel_bias (sf.M33): Derivative of Dv w.r.t. accel_bias evaluated at accel_bias_hat
+        Dv_D_gyro_bias (sf.M33): Derivative of Dv w.r.t. gyro_bias evaluated at gyro_bias_hat
+        accel_bias_hat (sf.V3): The bias of the accelerometer measurements used during pre-integration
+        gyro_bias_hat (sf.V3): The bias of the gyroscope measurements used during pre-integration
+        gravity (sf.V3): Acceleration due to gravity (in the same frame as pose_x and vel_x)
+        dt (T.Scalar): The time between timestep i and j: t_j - t_i
+        epsilon (T.Scalar): epsilon used for numerical stability
+    """
+    delta_gyro_bias = gyro_bias_i - gyro_bias_hat
+    delta_accel_bias = accel_bias_i - accel_bias_hat
+
+    # Correct preintegrated measurements for updated bias estimates
+    corrected_DR = DR * sf.Rot3.from_tangent(
+        (DR_D_gyro_bias * delta_gyro_bias).to_storage(), epsilon
+    )
+    corrected_Dv = Dv + Dv_D_gyro_bias * delta_gyro_bias + Dv_D_accel_bias * delta_accel_bias
+    corrected_Dp = Dp + Dp_D_gyro_bias * delta_gyro_bias + Dp_D_accel_bias * delta_accel_bias
+
+    res_R = (corrected_DR.inverse() * pose_i.R.inverse() * pose_j.R).to_tangent(epsilon)
+    res_v = pose_i.R.inverse() * (vel_j - vel_i - gravity * dt) - corrected_Dv
+    res_p = (
+        pose_i.R.inverse() * (pose_j.t - pose_i.t - vel_i * dt - sf.S(1) / 2 * gravity * dt ** 2)
+        - corrected_Dp
+    )
+
+    res = sf.V9(*res_R, *res_v, *res_p)
+
+    def upper_triangular_mul(ut: sf.M99, v: sf.V9) -> sf.V9:
+        """
+        Returns ut * v. Use to avoid expressions containing the zero-valued lower triangular
+        portion of ut.
+
+        Precondition: ut is upper triangular
+        """
+        return sf.V9([(ut[k, k:] * v[k:, 0])[0] for k in range(9)])
+
+    return upper_triangular_mul(sqrt_info, res)
+
+
+def right_jacobian(tangent: sf.V3, epsilon: T.Scalar) -> sf.M33:
+    """
+    The right jacobian J(v) is d/du Log(Exp(v)^{-1} Exp(v + u)), i.e., a matrix such that
+    Exp(v + dv) ~= Exp(v) Exp(J(v) dv), where v and u are tangent vectors of SO(3).
+
+    Returns J(tangent).
+    """
+    norm = tangent.norm(epsilon)
+    tangent_hat = sf.Matrix.skew_symmetric(tangent)
+    out = sf.M33.eye()
+    out += ((1 - sf.cos(norm)) / (norm ** 2)) * tangent_hat
+    out += ((norm - sf.sin(norm)) / (norm ** 3)) * (tangent_hat * tangent_hat)
+    return out
```

Comment:
Presumably this is less expensive than computing this jacobian automatically (does it even work to sub u=0 directly instead of taking the limit?), and the instability / precision loss of `(1 - cos x) / x**2` isn't a problem?

Reply:
I hadn't actually considered this carefully (whether to do it automatically or not) when I wrote it. I saw it in the paper, calculated it by hand to convince myself it was right, then typed it in here. Could look into calculating it symbolically. It looks an awful lot like the problem of `tangent_jacobians` with a few minor differences.

And `(1 - cos x) / x ** 2` and `x - sin(x) / x ** 3` does look quite concerning.

On paper it's no real worry as these are well behaved at 0 for the exact same reason `sinc` function is (they look similar too). Moreover, since we use epsilon when calculating `norm`, it will never actually be 0 (and we don't risk dividing my 0).

But in terms of floating point arithmetic, it worries me as it kind of reminds me of the problem of calculating `log(1 + x)` for small x. in this case, `cos(x)` will be roughly `1 - 0.5 * x^2`. If `x` is `sqrt(epsilon)`, then `cos(x)` will only be about 10 floating point numbers below 1, so we can expect (1 - cos(x)) to have a relative error of about 10%, Then when we divide by `x^2` (again, evaluating to epsilon), we'll be dividing by the correct value, and end up possibly with an answer of `0.55` or `0.45` instead of the correct `0.5`.

Okay, I just sat down in python and did this
``` python
from symforce import numeric_epsilon as epsilon
import math

def truncated_cos(x):
    y = math.sqrt(x ** 2 + epsilon)
    return (1 - math.cos(y)) / y ** 2

print(truncated_cos(1e-8))
print(truncated_cos(2e-8))
print(truncated_cos(3e-8))
```
They should all be slightly less than `0.5`, but you'll see that the value jumps around a bunch. `truncated_cos(1e-8)` gives `0.478` which is roughly the sort of error I expected from my back of the envelope math above.

And note, the value of `x` we use here is the norm of `corrected_gyro * dt`, so it's not crazy to think that we could end up passing a pretty small value here.

Reply:
Okay. We spoke about this in person, and decided the best course of action was to do `norm = tangent.norm(sf.sqrt(epsilon))`.

This way, `cos(norm)` will always be far enough from 1 to not loose too many digits of precision when we subtract it from 1, while not being such a large perturbation as to significantly affect the result anywhere (the derivative of these functions is bounded and pretty small everywhere, `< 1`).